vis_gen.py

```diff
@@ -0,0 +1,565 @@
+              .format(total_cp, start, end))
+        for cur_ord in range(ord_start, ord_end+1):
+            if (cur_ord - ord_start) % 50 == 0:
+                print("Now writing {}st code point."
+                      .format(str(cur_ord - ord_start + 1)))
+            code_point = chr(cur_ord)
+            self.visualize_single(code_point, False, x=x, y=y)
+        print("Finished.")
+        print("Images stored in directory {}.".format(out_dir_abs))
+
+        self.__check_out_dir = True
+
+
+    def visualize_single(self, code_point, check_out_dir=True, x=None, y=None):
+        """Write rendered text image for specific code point to output
+        directory, text is positioned at (x, y). If both x and y are set to None
+        (default), the text will be centered.
+
+        Args:
+            code_point: Str, single unicode code point
+            check_out_dir: Bool, set True if need to check output directory
+            x: Int, x coordinate of the position of text, in number of pixels.
+            y: Int, y coordinate of the position of text, in number of pixels.
+
+        Returns:
+            file_path: Str, path to file
+
+        Raises:
+            OSError: if specified directory cannot be created
+        """
+        # Get absolute directory path
+        out_dir_abs = os.path.join(os.getcwd(), self.out_dir)
+        if check_out_dir:
+            self.__check_create_out_dir_abs(out_dir_abs)
+
+        # Create and configure ImageSurface
+        figure_dimensions = Vector(self.image_size, self.image_size)
+        pix = qah.ImageSurface.create(format=self.__cairo_format,
+                                      dimensions=figure_dimensions)
+
+        # Create context
+        ctx = qah.Context.create(pix)
+
+        # Select anti-aliasing style in font options
+        font_options = ctx.font_options
+        font_options.antialias = self.__antialias
+        # Set color
+        ctx.set_source_colour(self.__canvas_color)
+        ctx.paint()
+        ctx.set_source_colour(self.__text_color)
+        # Set font face and size
+        ctx.set_font_face(self.__font_face)
+        ctx.set_font_size(self.__font_size)
+        ctx.set_font_options(font_options)
+
+        # Position text
+        if x is None and y is None:
+            self.__center_text(ctx, code_point)
+        elif x is None or y is None:
+            raise ValueError("Expect both x and y to be specified for "
+                             "alternative positioning.")
+        else:
+            self.__position_text(ctx, code_point, x, y)
+
+        # Get file name and file path
+        filename = self.__get_filename(code_point)
+        file_path = os.path.join(out_dir_abs, filename)
+
+        # Show text and write to file
+        ctx.show_text(code_point)
+        pix.flush()
+        pix.write_to_png(file_path)
+
+        return file_path
+
+    def train_test_split(self, num_test=100):
```

Comment:
This function has nothing to do with visual generation. Please move it into a separate file.